python-api/tests/integrationtests/baseclass.py

```diff
@@ -0,0 +1,412 @@
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG)
+        logging.basicConfig(
+            format="%(filename)s:%(lineno)d (%(funcName)s): %(message)s")
+
+        self.serv = None
+        self._sources = None
+        self.controller = None
+
+    def setup_server(self, record_file=False):
+        """Set up a gst-switch server for testing."""
+        assert self.serv is None
+
+        self.log.info("setting up Server")
+        self.serv = Server(path=PATH, video_format="debug",
+                           record_file=record_file)
+
+        self.log.info("running Server")
+        self.serv.run()
+        assert self.serv.is_alive()
+
+        self.log.info("waiting for Server to open Controller-Port")
+        self.serv.wait_for_output('tcp:host=0.0.0.0,port=5000')
+
+        self.log.info("setting up TestSources")
+        self._sources = TestSources(
+            video_port=self.serv.video_port,
+            audio_port=self.serv.audio_port)
+
+    def setup_controller(self):
+        """Create Controller object and call setup_controller."""
+        self.log.info("setting up Controller")
+        self.controller = Controller()
+
+        self.log.info("connecting Controller to Server")
+        self.controller.establish_connection()
+        self.serv.wait_for_output('registered: ')
+
+        assert self.controller.connection is not None
+
+    def new_test_video(self, pattern=None):
+        """Start a new Video-Testsource and wait until it's ready"""
+        self.serv.wait_for_output(':::3000')
+        self._sources.new_test_video(pattern=pattern)
+
+    def new_test_audio(self, freq=110, wave=AudioSrc.WAVE_SINE):
+        """Start a new Audio-Testsource and wait until it's ready"""
+        self.serv.wait_for_output(':::4000')
+        self._sources.new_test_audio(freq=freq, wave=wave)
+
+    def teardown_method(self, _):
+        """Tear down called automatically after every test_XXXX method."""
+        self.controller = None
+
+        # Kill all the sources
+        if self._sources is not None:
+            self.log.info("terminating Video-TestSource")
+            self._sources.terminate_video()
+
+            self.log.info("terminating Audio-TestSource")
+            self._sources.terminate_audio()
+        self._sources = None
+
+        if self.serv is not None:
+            self.log.info("terminating Server")
+            self.serv.terminate(cov=True)
+
+            if self.serv.proc:
+                poll = self.serv.proc.poll()
+                if poll == -11:
+                    self.log.error("Server exited with Segmentation Fault")
+                if poll != 0:
+                    self.log.error("Server exited Error Eode %s", poll)
+
+            log = open('server.log')
+            print(log.read())
+
+        self.serv = None
+
+
+class IntegrationTestbaseMainloop(IntegrationTestbase):
+    """Base class for integration tests that require a GLib-Mainloop.
```

Comment:
Can you give an example of what requires you to have a GLib-Mainloop? (So people know when they should be using this base)